File: iot/amqp/sections/AMQPValue.py
import iot.amqp.header.api.AMQPWrapper as AMQPWrapper
import iot.amqp.header.api.AMQPUnwrapper as AMQPUnwrapper
import iot.amqp.constructor.DescribedConstructor as DescribedConstructor
import iot.amqp.avps.SectionCode as SectionCode
import iot.amqp.tlv.impl.TLVFixed as TLVFixed
import iot.amqp.tlv.impl.TLVNull as TLVNull
import iot.amqp.avps.AMQPType as AMQPType

class amqpValue():
    def __init__(self, value):
        self.value = value

    def getValue(self):
        val = None
        wrapper = AMQPWrapper.amqpWrapper() 
        if self.value is not None:
            val = wrapper.wrap(self.value)
        else:
            val = TLVNull.tlvNull()

        if val is not None:
            constructor = DescribedConstructor.describedConstructor(val.getCode(), TLVFixed.tlvFixed(AMQPType.amqpType.getValueByKey('SMALL_ULONG'), 0x77))
            val.setConstructor(constructor)
        return val

    def fill(self, value):
        unwrapper = AMQPUnwrapper.amqpUnwrapper()
        if value is not None:
            self.value = unwrapper.unwrap(value)

    def getCode(self):
        return SectionCode.sectionCode.getValueByKey('VALUE')

    def toString(self):
        return 'AMQPValue [value=' + str(self.value) + ']'

    def getValue(self):
        return self.value

    def setValue(self, value):
        self.value = value

File: IoTbrokerClient/src/iot-protocols/amqp/classes/section/amqpvalue.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqpvalue.h"
#include "iot-protocols/amqp/classes/headerapi/amqpwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpunwrapper.h"
#include "iot-protocols/amqp/classes/constructors/amqpdescribedconstructor.h"

AMQPValue::AMQPValue()
{

}

TLVAMQP *AMQPValue::getValue()
{
    TLVAMQP *value = this->valueObject != NULL ? AMQPWrapper::wrapObject(this->valueObject) : new AMQPTLVNull();

    unsigned char * data = NULL;
    data = malloc(sizeof(char));
    data[0] = (char) 0x77;
    AMQPType *type = new AMQPType(AMQP_SMALL_ULONG_TYPE);
    AMQPTLVFixed *fixed = new AMQPTLVFixed(type, data);

    AMQPType *constructorType = new AMQPType(value->getTypeValue());
    AMQPDescribedConstructor *constructor = new AMQPDescribedConstructor(constructorType, fixed);

    value->setConstructor(constructor);
    return value;
}

AMQPSectionCode *AMQPValue::getCode()
{
    return new AMQPSectionCode(AMQP_VALUE_SECTION_CODE);
}

void AMQPValue::fill(TLVAMQP *value)
{
    if (value->isNull()) {
        this->valueObject = AMQPUnwrapper::unwrap(value);
    }
}

JsonVariant *AMQPValue::getValueObject() const
{
    return valueObject;
}

void AMQPValue::setValueObject(JsonVariant *value)
{
    valueObject = value;
}
